### PR TITLE
feat : add `complex/base/parse`

### PR DESCRIPTION
**Pull request by marsian83**
## Description

This PR tackles the parsing as discussed [here](https://github.com/stdlib-js/stdlib/issues/1370#issuecomment-1962745786)

This implements a parser which will convert a complex number represented as a string to a complex like object

This can be used as a base for `@stdlib/complex/assert/is-complex` and `@stdlib/complex/parse-32` and `@stdlib/complex/parse-64`. 
Detailed discussion as to why this approach may be chosen can be found #1332 & #1370

## Related Issues

#1332
#1333
#1370

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by kgryte** [comment]
@marsian83 Looks like linting is failing. Mind resolving those lint errors?

**Comment by marsian83** [comment]
> @marsian83 Looks like linting is failing. Mind resolving those lint errors?

While linting this line is behaving weirdly
image

It requires me to have no spaces in JSDOC type definitions but also requires me to have spaces after the curly bracket opening and before it closing. how do I resolve this conflicting requirement

**Comment by marsian83** [comment]
image

These three are proving to be quite problematic to me, the first one, I do not understand.
The second is the one I discussed above.
The third one is asking me to require `Number` which is an inbuilt JS module?

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+'use strict';
+
+// MAIN //
+
+/**
+* Parses a string representing a complex number into a complex-like object.
+*
+* @param {string} str - The string representing the complex number.
+* @returns { {re: number, im: number} } A complex-like object with real and imaginary parts.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js]:
```suggestion
* @returns {ComplexLike} an object containing real and imaginary parts
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ */
+
+'use strict';
+
+// MAIN //
+
+/**
+* Parses a string representing a complex number into a complex-like object.
+*
+* @param {string} str - The string representing the complex number.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js]:
```suggestion
* @param {string} str - input string
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * limitations under the License.
+ */
+
+'use strict';
+
+// MAIN //
+
+/**
+* Parses a string representing a complex number into a complex-like object.
+*
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js:31]:
```suggestion
*
* ## Notes
*
* -   The function returns an object containing the following properties:
*
*     -   **re**: real component
*     -   **im**: imaginary component
*
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+* @param {string} str - The string representing the complex number.
+* @returns { {re: number, im: number} } A complex-like object with real and imaginary parts.
+*
+* @example
+* var parse = require('@stdlib/complex/base/parse');
+*
+* var str = '4 + 6i';
+*
+* var z = parse(str);
+* // returns { 're' : 4, 'im' : 6 }
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js]:
```suggestion
* // returns { 're': 4, 'im': 6 }
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+*
+* @param {string} str - The string representing the complex number.
+* @returns { {re: number, im: number} } A complex-like object with real and imaginary parts.
+*
+* @example
+* var parse = require('@stdlib/complex/base/parse');
+*
+* var str = '4 + 6i';
+*
+* var z = parse(str);
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js]:
```suggestion
* var z = parse( str );
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/main.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+* Parses a string representing a complex number into a complex-like object.
+*
+* @param {string} str - The string representing the complex number.
+* @returns { {re: number, im: number} } A complex-like object with real and imaginary parts.
+*
+* @example
+* var parse = require('@stdlib/complex/base/parse');
+*
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/main.js]:
```suggestion
```

File: lib/node_modules/@stdlib/complex/base/parse/lib/index.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+* limitations under the License.
+*/
+
+'use strict';
+
+/**
+* Parses a string representing a complex number into a complex like object.
+*
+* @param {string} str
+* @returns {(*|{ re : number, im : number })} complex like object
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/index.js]:
```suggestion
* @module @stdlib/complex/base/parse
```

See how other packages do JSDoc.

File: lib/node_modules/@stdlib/complex/base/parse/lib/index.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+*
+* @example
+* var parse = require( '@stdlib/complex/base/parse' );
+*
+* var str = '4 + 6i';
+*
+* var z = parse( str );
+* // returns { re : 4, im : 6 }
```
Comment (kgryte) [inline lib/node_modules/@stdlib/complex/base/parse/lib/index.js]:
```suggestion
* // returns { 're': 4, 'im': 6 }
```

**Comment by marsian83** [comment]
I have noted and understood the changes you made and also commited and signed them off
I am still confused about the linter asking me to `require` 'Number' though

**Comment by marsian83** [comment]
There was a slight lint error in the license headers. I have fixed and pushed changes

File: lib/node_modules/@stdlib/complex/base/parse/docs/types/test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+import parse = require( './index' );
+
+
+// TESTS //
+
+// The function returns a complex like object...
+{
+	const str = '54 + 3i';
+	parse( str ); // $ExpectType any
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/docs/types/test.ts]:
```suggestion
	parse( str ); // $ExpectType {re: number; im: number} | null
```

File: lib/node_modules/@stdlib/complex/base/parse/docs/repl.txt [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    Parses a string representing a complex number into a complex like object.
+
+    Parameters
+    ----------
+    str: string
+
+    Returns
+    -------
+    out: {re: number, im: number} | null
+        null or Complex like object.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/docs/repl.txt]:
```suggestion
        Complex like object or null.
```

File: lib/node_modules/@stdlib/complex/base/parse/docs/repl.txt [hunk 11]
```diff
@@ -0,0 +1,20 @@
+{{alias}}( str )
+    Parses a string representing a complex number into a complex like object.
+
+    Parameters
+    ----------
+    str: string
+
+    Returns
+    -------
+    out: {re: number, im: number} | null
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/docs/repl.txt]:
```suggestion
    out: ComplexLike | null
```

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+* limitations under the License.
+*/
+
+'use strict';
+
+var parse = require( '@stdlib/complex/base/parse/lib' );
+
+var str = '5 + 4i';
+var z = parse( str );
+// z = '{"re":5,"im":4}'
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js]:
```suggestion
// => '{"re":5,"im":4}'
```

**Review by Planeshifter — APPROVED** [review]

**Comment by Planeshifter** [comment]
@marsian83 Thanks for making the requested changes and working on this new package in the first place! Made some further small changes to align with project conventions and will merge once CI passes. 🚀

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+str = 'Infinity + 2.34i';
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => '{"re":Infinity,"im":2.34}'
+
+str = 'NaN + i4';
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => '{"re":NaN,"im":4}'
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js]:
```suggestion
// => 'null'
```

**Comment by marsian83** [comment]
> @marsian83 Thanks for making the requested changes and working on this new package in the first place! Made some further small changes to align with project conventions and will merge once CI passes. 🚀

Thanks for bringing to my attention the conventions I missed, I will keep these in mind from next time on now that you have pointed these out.

The reason the CI is failing is quite puzzling to me, The linter is asking me to `require` the Javascript global function `Number`, the one which we use to convert a string to number. How can I require a global function ?
What is the way to fix this linting error?

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+str = '45i55 + 5';
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => null
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js:41]:
```suggestion
// => 'null'
```

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+str = '5 + 6 + 10e4i';
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => '{"re":11,"im":10000}'
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js]:
```suggestion
// => '{"re":11,"im":100000}'
```

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+str = {};
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => null
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js:51]:
```suggestion
// => 'null'
```

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+* limitations under the License.
+*/
+
+'use strict';
+
+var parse = require( '@stdlib/complex/base/parse/lib' );
+
+var str = '5 + 4i';
+var z = parse( str );
+// => '{"re":5,"im":4}'
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js:26]:
```suggestion
// returns '{"re":5,"im":4}'
```

File: lib/node_modules/@stdlib/complex/base/parse/examples/index.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+str = 'Infinity + 2.34i';
+z = parse( str );
+console.log( JSON.stringify( z ) );
+// => '{"re":Infinity,"im":2.34}'
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/complex/base/parse/examples/index.js]:
```suggestion
// => '{"re":null,"im":2.34}'
```

**Comment by marsian83** [comment]
@Planeshifter  Thanks 😄  for bringing to my attention the conventions I missed, I will keep these in mind from next time on now that you have pointed these out.

**Pull request merged**